File: nreduce/benchmark/mandelbrot.pl
#!/usr/bin/perl -w

sub mandel
{
  my ($Cr,$Ci,$iterations) = @_;
  my $Zr = 0.0;
  my $Zi = 0.0;
  for (my $r = 0; $r < $iterations; $r++) {
    $newZr = ($Zr*$Zr) - ($Zi*$Zi) + $Cr;
    $newZi = (2.0*($Zr*$Zi)) + $Ci;
    $mag = sqrt(($newZr*$newZr)+($newZi*$newZi));
    if ($mag > 2.0) {
      return $r;
    }
    $Zr = $newZr;
    $Zi = $newZi;
  }
  return $iterations;
}


sub printcell
{
  my $num = $_[0];
  if ($num > 40) {
    print "  ";
  }
  elsif ($num > 30) {
    print "''";
  }
  elsif ($num > 20) {
    print "--";
  }
  elsif ($num > 10) {
    print "**";
  }
  else {
    print "##";
  }
}

sub mloop
{
  my ($minx,$maxx,$xincr,$miny,$maxy,$yincr) = @_;
  $maxx += $xincr;
  for ($y = $miny; $y <= $maxy; $y += $yincr) {
    for ($x = $minx; $x <= $maxx; $x += $xincr) {
      printcell(mandel($x,$y,4096));
    }
    print "\n";
  }
}

$incr = 0.01;
if (0 <= $#ARGV) {
  $incr = $ARGV[0];
}
print "incr = $incr\n";
mloop(-1.5,0.5,$incr,-1.0,1.0,$incr);

File: evaluation/benchmark/mandelbrot.java
public class mandelbrot
{
  static double mandel(double Cr, double Ci)
  {
    double res;
    double Zr = 0.0;
    double Zi = 0.0;
    for (res = 0; res < 4096; res++) {
      double newZr = ((Zr * Zr) - (Zi * Zi)) + Cr;
      double newZi = (2.0 * (Zr * Zi)) + Ci;
      double mag = Math.sqrt((newZr * newZr) + (newZi * newZi));
      if (mag > 2.0)
        return res;
      Zr = newZr;
      Zi = newZi;
    }
    return res;
  }

  static void printcell(double num)
  {
    if (num > 40)
      System.out.print("  ");
    else if (num > 30)
      System.out.print("''");
    else if (num > 20)
      System.out.print("--");
    else if (num > 10)
      System.out.print("**");
    else
      System.out.print("##");
  }

  static void mloop(double minx, double maxx, double miny, double maxy, double incr)
  {
    double x;
    double y;
    for (y = miny; y < maxy; y += incr) {
      for (x = minx; x < maxx; x += incr)
        printcell(mandel(x,y));
      System.out.println();
    }
  }

  public static void main(String[] args)
  {
    int n = 32;
    if (args.length > 0)
      n = Integer.parseInt(args[0]);

    double incr = 2.0/((double)n);
    mloop(-1.5,0.5,-1.0,1.0,incr);
  }
}
